python/ray/data/_internal/execution/operators/map_operator_state.py

```diff
@@ -183,12 +180,32 @@ def _create_task(self, bundle: RefBundle) -> None:
         ref: ObjectRef[ObjectRefGenerator] = self._task_submitter.submit(input_blocks)
         task = _TaskState(bundle)
         self._tasks[ref] = task
-        self._tasks_by_output_order[self._next_task_index] = task
-        self._next_task_index += 1
+        self._output_queue.add_pending_task(task)
         self._obj_store_mem_cur += bundle.size_bytes()
         if self._obj_store_mem_cur > self._obj_store_mem_peak:
             self._obj_store_mem_peak = self._obj_store_mem_cur
 
+    def current_resource_usage(self) -> ExecutionResources:
+        if isinstance(self._task_submitter, ActorPoolSubmitter):
+            num_active_workers = self._task_submitter._actor_pool.size()
+        else:
+            num_active_workers = self.num_active_work_refs()
+        return ExecutionResources(
+            cpu=self._ray_remote_args.get("num_cpus", 0) * num_active_workers,
+            gpu=self._ray_remote_args.get("num_gpus", 0) * num_active_workers,
+            object_store_memory=self._obj_store_mem_cur,
+        )
+
+    def incremental_resource_usage(self) -> ExecutionResources:
+        if isinstance(self._task_submitter, ActorPoolSubmitter):
+            # TODO(ekl) this should be non-zero if all actors are saturated, otherwise
+            # we will submit way too many tasks.
+            return ExecutionResources(cpu=0, gpu=0)
+        return ExecutionResources(
+            cpu=self._ray_remote_args.get("num_cpus", 0),
+            gpu=self._ray_remote_args.get("num_gpus", 0),
+        )
```

Comment:
Same here, could we push this down into the submitters? Although I'm not sure how to handle `num_active_work_refs`... This can probably all wait until a future refactor.

Reply:
This one is trickier, I'll leave it for the refactor.